File: scala/pattern_match.scala
trait UOM
case object centimeter extends UOM
case object inch extends UOM

// basic pattern match
def convert_length(value: (UOM, Double)): (UOM, Double) = {
  val (u, d) = value
  u match {
    case centimeter => (u, (d / 2.54))
    case inch => (u, (d * 2.54))
  }
}

// // with guards
// def convert_length(value: (UOM, Double)): (UOM, Double) = {
//   value match {
//     case (u:UOM, d: Double) if u == centimeter => (u, (d / 2.54))
//     case (u:UOM, d: Double) if u == inch => (u, (d * 2.54))
//   }
// }

File: erlang/pattern_match.erl
-module (pattern_match).
-export ([convert_length/1]).
% Atoms are lower case and can't have a value

convert_length({centimeter, X}) ->
  {inch, X / 2.54};
convert_length({inch, Y}) ->
  {centimeter, Y * 2.54}.
% convert_length({inch, Y}) ->
%   {centimeter, Y * 2.54};
% convert_length({millimeter, Y}) ->
%   {inch, Y / 25.4}.
